app/controllers/users_controller.rb

```diff
@@ -1,4 +1,52 @@
 class UsersController < ApplicationController
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
 
+  get '/signup' do
+    if !logged_in?
+      erb :'users/create_user'
+    else
+      redirect '/tweets'
+    end
+  end
 
+  get '/login' do
+    if !logged_in?
+      erb :'users/login'
+    else
+      redirect to '/tweets'
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params[:username])
+    if @user != nil && @user.authenticate(params[:password])
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/signup'
+    end
+  end
+
+  post '/signup' do
+    if !params[:username].empty? && !params[:email].empty? && !params[:password].empty?
```

Comment:
I think you could use [ActiveRecord validations](http://guides.rubyonrails.org/active_record_validations.html) (e.g. `validates_presence_of :username`) to perform these types of checks more easily.